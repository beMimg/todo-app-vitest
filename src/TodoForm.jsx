import { useState } from "react";

export function TodoForm({ onSubmit }) {
  const [inputValue, setInputValue] = useState("");

  function handleSubmit(e) {
    e.preventDefault();
    onSubmit(inputValue);
    setInputValue("");
  }

  return (
    <form onSubmit={handleSubmit}>
      <label>Todo name:</label>
      <input
        type="text"
        value={inputValue}
        onChange={(e) => setInputValue(e.target.value)}
      />
      <button type="submit">Submit</button>
    </form>
  );
}
