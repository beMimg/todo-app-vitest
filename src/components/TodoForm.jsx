import { useState } from "react";

export function TodoForm({ onSubmit }) {
  const [inputValue, setInputValue] = useState("");

  function handleSubmit(e) {
    e.preventDefault();
    onSubmit(inputValue);
    setInputValue("");
  }

  return (
    <form className="todo-form" onSubmit={handleSubmit}>
      <label htmlFor="todo-name">Todo name:</label>
      <input
        type="text"
        value={inputValue}
        id="todo-name"
        name="todo-name"
        data-testid="input-element"
        onChange={(e) => setInputValue(e.target.value)}
      />
      <button type="submit">Submit</button>
    </form>
  );
}
