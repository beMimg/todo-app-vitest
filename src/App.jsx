import { useState } from "react";
import { TodoForm } from "./TodoForm";

function App() {
  const [todos, setTodos] = useState([]);

  function handleTodoSubmit(todoName) {
    setTodos((todos) => [
      ...todos,
      { name: todoName, id: crypto.randomUUID(), time: new Date() },
    ]);
  }

  return (
    <>
      <TodoForm onSubmit={handleTodoSubmit}></TodoForm>
      <ul data-testid="todos-list">
        {todos.map((todo) => (
          <li key={todo.id}>{todo.name}</li>
        ))}
      </ul>
    </>
  );
}

export default App;
