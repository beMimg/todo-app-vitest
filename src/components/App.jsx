import { useState } from "react";
import { TodoForm } from "./TodoForm";
import { EditForm } from "./EditForm";
import { TodoDisplay } from "./TodoDisplay";
import { TestsModal } from "./TestsModal";

function App() {
  const [todos, setTodos] = useState([]);
  const [editTodoName, setEditTodoName] = useState();
  const [displayModal, setDisplayModal] = useState(false);

  function handleTodoSubmit(todoName) {
    setTodos((todos) => [
      ...todos,
      {
        name: todoName,
        id: crypto.randomUUID(),
        time: new Date(),
        isBeingEdited: false,
      },
    ]);
  }

  function handleDelete(id) {
    const filtredArray = todos.filter((todo) => todo.id !== id);
    setTodos(filtredArray);
  }

  function turnEditOn(id) {
    if (todos.some((todo) => todo.isBeingEdited)) {
      return;
    } else {
      setTodos((prevTodos) => {
        return prevTodos.map((todo) => {
          if (todo.id === id) {
            return { ...todo, isBeingEdited: true };
          } else {
            return todo;
          }
        });
      });
    }
  }

  function changeTodoName(e, id) {
    e.preventDefault();
    setTodos((prevTodos) => {
      return prevTodos.map((todo) => {
        if (todo.id === id) {
          return { ...todo, name: editTodoName, isBeingEdited: false };
        } else {
          return todo;
        }
      });
    });
    setEditTodoName("");
  }

  function cancelEdit(id) {
    setTodos((prevTodos) => {
      return prevTodos.map((todo) => {
        if (todo.id === id) {
          return { ...todo, isBeingEdited: false };
        } else {
          return todo;
        }
      });
    });
    setEditTodoName("");
  }

  function handleCloseModal() {
    setDisplayModal(false);
  }
  const isThereAnActiveEdit = todos.some((todo) => todo.isBeingEdited);

  return (
    <>
      <button
        className="tests-btn"
        onClick={() => {
          setDisplayModal(true);
        }}
      >
        Tests
      </button>
      {displayModal && <TestsModal onClose={handleCloseModal}></TestsModal>}
      <TodoForm onSubmit={handleTodoSubmit}></TodoForm>
      <div data-testid="todos-list" className="todos-list">
        {todos.map((todo) =>
          todo.isBeingEdited ? (
            <EditForm
              onSubmit={changeTodoName}
              onChange={setEditTodoName}
              onClick={cancelEdit}
              todo={todo}
              editTodoName={editTodoName}
              key={todo.id}
            ></EditForm>
          ) : (
            <TodoDisplay
              todo={todo}
              onDelete={handleDelete}
              onEdit={turnEditOn}
              isAvailable={isThereAnActiveEdit}
              key={todo.id}
            ></TodoDisplay>
          )
        )}
      </div>
    </>
  );
}

export default App;
