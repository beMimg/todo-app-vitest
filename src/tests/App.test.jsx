import { describe, it, expect, vi } from "vitest";
import {
  fireEvent,
  queryByText,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import App from "../App";
import userEvent from "@testing-library/user-event";
import { TodoForm } from "../TodoForm";

describe("App Component", () => {
  it("renders app", () => {
    const { container } = render(<App></App>);
    expect(container).toMatchSnapshot();
  });

  it("It displays todo when submited", async () => {
    const user = userEvent.setup();

    render(<App></App>);

    const button = screen.getByRole("button", { name: "Submit" });

    const inputElement = screen.getByTestId("input-element");

    fireEvent.change(inputElement, { target: { value: "New Todo" } });

    expect(screen.queryByText("New Todo")).toBeNull();

    user.click(button);

    await screen.findByText("New Todo");
    await waitFor(() => {
      expect(screen.queryByText("dsad")).toBeNull();
    });
  });
});
