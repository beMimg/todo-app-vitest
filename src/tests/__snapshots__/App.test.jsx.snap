// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App Component > renders app 1`] = `
<div>
  <form>
    <label
      for="todo-name"
    >
      Todo name:
    </label>
    <input
      data-testid="input-element"
      id="todo-name"
      name="todo-name"
      type="text"
      value=""
    />
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
  <ul
    data-testid="todos-list"
  />
</div>
`;

exports[`App Component > renders form 1`] = `
<div>
  <form>
    <label
      for="todo-name"
    >
      Todo name:
    </label>
    <input
      data-testid="input-element"
      id="todo-name"
      name="todo-name"
      type="text"
      value=""
    />
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
  <ul
    data-testid="todos-list"
  />
</div>
`;
